src/servers/tracer.cc

```diff
@@ -242,4 +240,203 @@ TraceManager::TraceActivity(
       << "\",\"ns\":" << timestamp_ns << "}]}";
 }
 
+void
+TraceManager::TraceTensorActivity(
+    TRITONSERVER_InferenceTrace* trace,
+    TRITONSERVER_InferenceTraceActivity activity, const char* name,
+    TRITONSERVER_DataType datatype, const void* base, size_t byte_size,
+    const int64_t* shape, uint64_t dim_count,
+    TRITONSERVER_MemoryType memory_type, int64_t memory_type_id, void* userp)
+{
+  if ((activity != TRITONSERVER_TRACE_TENSOR_INPUT) &&
+      (activity != TRITONSERVER_TRACE_TENSOR_OUTPUT)) {
+    LOG_ERROR << "Unsupported activity: "
+              << TRITONSERVER_InferenceTraceActivityString(activity);
+    return;
+  }
+
+  if ((memory_type != TRITONSERVER_MEMORY_CPU) &&
+      (memory_type != TRITONSERVER_MEMORY_CPU_PINNED)) {
+    LOG_ERROR << "Unsupported memory type: "
+              << TRITONSERVER_MemoryTypeString(memory_type);
+    return;
+  }
+
+  void* buffer_base = const_cast<void*>(base);
+
+  uint64_t id;
+  LOG_TRITONSERVER_ERROR(
+      TRITONSERVER_InferenceTraceId(trace, &id), "getting trace id");
+
+  // The function may be called with different traces but the same 'userp',
+  // group the activity of the same trace together for more readable output.
+  auto ts = reinterpret_cast<TraceStreams*>(userp);
+  std::stringstream* ss = nullptr;
+  {
+    if (activity == TRITONSERVER_TRACE_REQUEST_START) {
+      std::unique_ptr<std::stringstream> stream(new std::stringstream());
+      ss = stream.get();
+      std::lock_guard<std::mutex> lk(ts->mtx_);
+      ts->streams_.emplace(id, std::move(stream));
+    } else {
+      std::lock_guard<std::mutex> lk(ts->mtx_);
+      ss = ts->streams_[id].get();
+    }
+  }
+
+  std::stringstream ss_tmp;
+
+  // collect and serialize trace details.
+  ss_tmp << ",{\"id\":" << id << ",\"activity\":\""
```

Comment:
Why not write directly to 'ss'... doing it this way introduces unnecessary copies.